File: crates/crucible-web/src/routes/fs.rs
//! File-tree explorer routes: `GET /api/fs/list` (project dir listing proxy),
//! `POST /api/fs/move` (DnD move / rename), `POST /api/fs/mkdir`,
//! `POST /api/fs/trash` (context-menu mutations), and `GET /api/fs/events`
//! (live filesystem-change SSE).

use crate::fs_events::FsEvent;
use crate::routes::helpers::{stream_version_frame, versioned};
use crate::routes::session::daemon_shape;
use crate::services::daemon::AppState;
use crate::{error::WebResultExt, WebError};
use axum::{
    extract::{Query, State},
    response::sse::{Event, KeepAlive, Sse},
    Json,
};
// The daemon owns every shape these four routes forward. Its module used to
// say so in a comment — "byte-identical to the TypeScript `FsEntry`" — which
// is the kind of contract that cannot fail a build. The routes read the types
// instead.
use crucible_daemon::{FsListing, FsMoveReply, FsTrashReply};
use futures::stream::Stream;
use serde::{Deserialize, Serialize};
use std::convert::Infallible;
use tokio_stream::StreamExt;
use utoipa::{IntoParams, ToSchema};
use utoipa_axum::{router::OpenApiRouter, routes};

pub fn fs_routes() -> OpenApiRouter<AppState> {
    OpenApiRouter::new()
        .routes(routes!(list_dir))
        .routes(routes!(move_path))
        .routes(routes!(mkdir_path))
        .routes(routes!(trash_path))
        .routes(routes!(fs_event_stream))
}

/// The two roots a mutation may name, and the allowlist each one selects.
///
/// This describes the wire; the field that carries it stays a `String` so that
/// an unknown kind is refused by the daemon, in the daemon's own sentence,
/// rather than by a body rejection that says only "unknown variant".
/// `every_root_kind_is_one_the_daemon_admits` walks an exhaustive match, so a
/// third kind cannot reach the document without a decision about it.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "lowercase")]
enum FsRootKind {
    /// A registered project, or a session's own workspace folder.
    Project,
    /// A registered kiln.
    Kiln,
}

#[derive(Debug, Deserialize, IntoParams)]
#[into_params(parameter_in = Query)]
struct FsListQuery {
    /// Absolute path of the root to list inside.
    root: String,
    /// Root-relative POSIX path of the directory. Empty lists the root.
    #[serde(default)]
    rel_path: String,
    /// Include entries git ignores. The file tree sends `true`.
    #[serde(default)]
    show_ignored: bool,
    /// Include dotfiles. `.git` never lists, whatever this says.
    #[serde(default)]
    show_hidden: bool,
}

/// `GET /api/fs/list` — one directory level inside a registered root.
///
/// All security (registry allowlist, path containment, symlink/dotfile
/// handling) is enforced daemon-side; this handler is a thin passthrough of
/// the daemon's listing envelope.
#[utoipa::path(
    get,
    path = "/api/fs/list",
    params(FsListQuery),
    responses(
        (status = 200, body = FsListing),
        (status = 422, description = "The daemon refuses the root or the relative path, and says why"),
        (status = 502, description = "The daemon could not list the directory, or answered a shape this route cannot read"),
    )
)]
async fn list_dir(
    State(state): State<AppState>,
    Query(query): Query<FsListQuery>,
) -> Result<Json<FsListing>, WebError> {
    let listing = state
        .daemon
        .fs_list_dir(
            &query.root,
            &query.rel_path,
            query.show_ignored,
            query.show_hidden,
        )
        .await
        .daemon_err()?;

    Ok(Json(daemon_shape(listing, "fs.list_dir")?))
}

#[derive(Debug, Deserialize, ToSchema)]
struct FsMoveBody {
    /// Absolute path of the root both ends sit inside.
    root: String,
    /// Which allowlist the daemon checks `root` against.
    #[schema(value_type = FsRootKind)]
    kind: String,
    /// Root-relative POSIX path of the entry to move.
    from_rel: String,
    /// Root-relative POSIX path it takes.
    to_rel: String,
}

/// `POST /api/fs/move` — move or rename one entry within one root.
///
/// The file-tree drag-and-drop backend. All security (allowlist, containment,
/// overwrite refusal) is daemon-side; this handler is a thin passthrough. Kiln
/// note and canvas moves carry the link-rewrite report, so the tree can tell
/// the user what happened to their links.
#[utoipa::path(
    post,
    path = "/api/fs/move",
    request_body = FsMoveBody,
    responses(
        (status = 200, body = FsMoveReply),
        (status = 422, description = "The daemon refuses the root, either path, or an overwrite, and says why"),
        (status = 502, description = "The daemon could not move the entry, or answered a shape this route cannot read"),
    )
)]
async fn move_path(
    State(state): State<AppState>,
    Json(body): Json<FsMoveBody>,
) -> Result<Json<FsMoveReply>, WebError> {
    let outcome = state
        .daemon
        .fs_move(&body.root, &body.kind, &body.from_rel, &body.to_rel)
        .await
        .daemon_err()?;
    Ok(Json(daemon_shape(outcome, "fs.move")?))
}

#[derive(Debug, Deserialize, ToSchema)]
struct FsPathBody {
    /// Absolute path of the root the entry sits inside.
    root: String,
    /// Which allowlist the daemon checks `root` against.
    #[schema(value_type = FsRootKind)]
    kind: String,
    /// Root-relative POSIX path of the entry.
    rel_path: String,
}

/// What `POST /api/fs/mkdir` answers.
///
/// The route's own shape, not the daemon's: `fs_mkdir` reports success as
/// `()`, so there is no daemon body to forward.
#[derive(Debug, Serialize, Deserialize, ToSchema)]
struct FsMkdirResponse {
    /// Always true. A refusal is an error status, not a `false`.
    created: bool,
}

/// `POST /api/fs/mkdir` — create a folder inside one root.
///
/// The tree's "New folder". Missing parents are created too. Thin daemon proxy.
#[utoipa::path(
    post,
    path = "/api/fs/mkdir",
    request_body = FsPathBody,
    responses(
        (status = 200, body = FsMkdirResponse),
        (status = 422, description = "The daemon refuses the root or the relative path, and says why"),
        (status = 502, description = "The daemon could not create the folder"),
    )
)]
async fn mkdir_path(
    State(state): State<AppState>,
    Json(body): Json<FsPathBody>,
) -> Result<Json<FsMkdirResponse>, WebError> {
    state
        .daemon
        .fs_mkdir(&body.root, &body.kind, &body.rel_path)
        .await
        .daemon_err()?;
    Ok(Json(FsMkdirResponse { created: true }))
}

/// `POST /api/fs/trash` — move one entry to the root's `.crucible/trash/`.
///
/// The tree's "Delete". Thin daemon proxy; kiln notes leave the index inline,
/// so backlinks re-resolve at once.
#[utoipa::path(
    post,
    path = "/api/fs/trash",
    request_body = FsPathBody,
    responses(
        (status = 200, body = FsTrashReply),
        (status = 422, description = "The daemon refuses the root or the relative path, and says why"),
        (status = 502, description = "The daemon could not trash the entry, or answered a shape this route cannot read"),
    )
)]
async fn trash_path(
    State(state): State<AppState>,
    Json(body): Json<FsPathBody>,
) -> Result<Json<FsTrashReply>, WebError> {
    let outcome = state
        .daemon
        .fs_trash(&body.root, &body.kind, &body.rel_path)
        .await
        .daemon_err()?;
    Ok(Json(daemon_shape(outcome, "fs.trash")?))
}

/// Live filesystem-change stream for the file-tree explorer.
///
/// The body schema describes one SSE `data:` payload, not the whole stream.
#[utoipa::path(
    get,
    path = "/api/fs/events",
    responses((
        status = 200,
        content_type = "text/event-stream",
        body = FsEvent,
        headers((
            "X-Crucible-Stream-Version" = u64,
            description = "The stream protocol this build speaks (also the first \
                           `stream_version` frame, for clients whose transport \
                           cannot read headers)"
        ))
    ))
)]
async fn fs_event_stream(
    State(state): State<AppState>,
) -> Result<
    (
        [(axum::http::HeaderName, String); 1],
        Sse<impl Stream<Item = Result<Event, Infallible>>>,
    ),
    WebError,
> {
    // ORDERING IS LOAD-BEARING: open the LOCAL broker channel BEFORE telling the
    // daemon to forward. The daemon only forwards "system" events after
    // `subscribe_sticky` lands; `EventBroker::dispatch` drops events for session
    // ids with no local subscriber (verified by
    // `dispatch_ignores_unsubscribed_sessions`). If we subscribed the daemon
    // first, an event forwarded in the window before `events.subscribe` created
    // the "system" channel would be dropped (a first-connection loss window).
    // Creating the broker channel first means every forwarded event has a buffer
    // to land in (broadcast buffers up to capacity regardless of SSE polling).
    let rx = state.events.subscribe("system").await;
    // Sticky: survives reconnect, shared by all browser connections.
    state.daemon.subscribe_sticky("system").await.daemon_err()?;

    let stream = futures::stream::iter([Ok(stream_version_frame())]).chain(
        tokio_stream::wrappers::BroadcastStream::new(rx)
            .filter_map(|result| result.ok())
            .filter_map(|event| {
                FsEvent::from_daemon_event(&event).map(|fe| {
                    let name = fe.event_name();
                    let data = serde_json::to_string(&fe).unwrap_or_default();
                    Ok(Event::default().event(name).data(data))
                })
            }),
    );

    Ok(versioned(Sse::new(stream).keep_alive(KeepAlive::default())))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_support::{
        mock_fs_listing, mock_fs_move_reply, mock_fs_trash_reply, request_json, shape, survives,
    };
    use crucible_daemon::{SkipReason, SkippedRef};
    use serde_json::json;

    /// The fixture, as the daemon put it on the wire.
    fn sent<T: serde::Serialize>(value: T) -> serde_json::Value {
        serde_json::to_value(value).expect("the daemon's type writes JSON")
    }

    /// The listing reaches the browser as the daemon wrote it.
    ///
    /// A round trip, not a status check: the mock answers with the daemon's
    /// own `FsListing`, the route reads it and writes it again, and the body
    /// has to be the same object. A route that dropped `truncated`, or a row
    /// that dropped `modified`, fails here.
    #[tokio::test]
    async fn list_dir_answers_the_declared_shape() {
        let listing = mock_fs_listing();
        let answered: FsListing = shape("GET", "/api/fs/list?root=/tmp/proj", None).await;

        assert_eq!(sent(&answered), sent(&listing));
        assert_eq!(answered.entries.len(), 2);
        assert!(answered.entries[0].is_dir);
        assert!(!answered.truncated);
    }

    /// A row keeps the two keys a reader could mistake for absent.
    ///
    /// `modified` is null for a platform that cannot report it and `status` is
    /// null until the git decoration lands. Both are WRITTEN, so a client
    /// reads null rather than `undefined`.
    #[tokio::test]
    async fn a_listing_row_writes_its_null_fields() {
        let answered: serde_json::Value = shape("GET", "/api/fs/list?root=/tmp/proj", None).await;

        let row = &answered["entries"][1];
        assert!(row.get("modified").is_some(), "modified is absent: {row}");
        assert_eq!(row["modified"], json!(null));
        assert!(row.get("status").is_some(), "status is absent: {row}");
        assert_eq!(row["status"], json!(null));
    }

    /// The link-rewrite report reaches the browser as the daemon wrote it.
    ///
    /// The second round trip. `rewritten_sources` and `skipped` are what the
    /// file tree tells the user about their links, and a named reply is the
    /// first thing that could drop them.
    #[tokio::test]
    async fn move_path_answers_the_declared_shape() {
        let reply = mock_fs_move_reply();
        let answered: FsMoveReply = shape(
            "POST",
            "/api/fs/move",
            Some(json!({
                "root": "/tmp/k",
                "kind": "kiln",
                "from_rel": "a.md",
                "to_rel": "notes/a.md",
            })),
        )
        .await;

        assert_eq!(sent(&answered), sent(&reply));
        assert!(answered.moved);
        assert_eq!(
            answered.skipped.as_ref().expect("a skipped list")[0].reason,
            SkipReason::Ambiguous
        );
    }

    /// Both arms of the move reply survive the round trip, each as itself.
    ///
    /// A move the link index does not watch answers `moved` ALONE. Writing the
    /// two report keys as null instead would tell the browser that a directory
    /// move rewrote no links, where today it says nothing about links at all.
    #[test]
    fn a_move_reply_omits_the_link_report_it_has_nothing_to_say_about() {
        let plain = FsMoveReply {
            moved: true,
            rewritten_sources: None,
            skipped: None,
        };
        assert_eq!(sent(&plain), json!({ "moved": true }));
        survives::<FsMoveReply>(&plain);

        let reported = FsMoveReply {
            moved: true,
            rewritten_sources: Some(vec!["index.md".to_string()]),
            skipped: Some(vec![SkippedRef {
                source_path: "other.md".to_string(),
                raw_target: "a".to_string(),
                reason: SkipReason::StaleSpan,
            }]),
        };
        assert_eq!(
            sent(&reported),
            json!({
                "moved": true,
                "rewritten_sources": ["index.md"],
                "skipped": [{
                    "source_path": "other.md",
                    "raw_target": "a",
                    "reason": "stale-span",
                }],
            })
        );
        survives::<FsMoveReply>(&reported);
    }

    #[tokio::test]
    async fn mkdir_path_answers_the_declared_shape() {
        let answered: FsMkdirResponse = shape(
            "POST",
            "/api/fs/mkdir",
            Some(json!({ "root": "/tmp/proj", "kind": "project", "rel_path": "new/dir" })),
        )
        .await;

        assert!(answered.created);
    }

    /// The trash reply reaches the browser as the daemon wrote it.
    #[tokio::test]
    async fn trash_path_answers_the_declared_shape() {
        let reply = mock_fs_trash_reply();
        let answered: FsTrashReply = shape(
            "POST",
            "/api/fs/trash",
            Some(json!({ "root": "/tmp/k", "kind": "kiln", "rel_path": "a.md" })),
        )
        .await;

        assert_eq!(sent(&answered), sent(&reply));
        assert!(answered.trashed);
        assert!(answered.trash_path.starts_with(".crucible/trash/"));
    }

    /// An unknown kind reaches the daemon rather than a body rejection.
    ///
    /// The document narrows `kind` to two values; the field stays a `String`
    /// so that this stays true. Typing the field as the enum would answer
    /// "unknown variant `folder`" in axum's own plain text, where the daemon
    /// names the two kinds it admits in the one error envelope every route
    /// uses. The mock admits every kind, so reaching it reads as a 200.
    #[tokio::test]
    async fn an_unknown_root_kind_reaches_the_daemon() {
        let (status, body) = request_json(
            "POST",
            "/api/fs/mkdir",
            Some(json!({ "root": "/tmp/proj", "kind": "folder", "rel_path": "x" })),
        )
        .await;

        assert_eq!(status, axum::http::StatusCode::OK, "{body}");
    }

    /// `FsRootKind`'s wire spelling, as one exhaustive table.
    ///
    /// A third kind fails to compile here until somebody decides what the
    /// daemon calls it.
    fn root_kind_on_the_wire(kind: FsRootKind) -> &'static str {
        match kind {
            FsRootKind::Project => "project",
            FsRootKind::Kiln => "kiln",
        }
    }

    /// The document's `kind` values are the ones `resolve_root` admits.
    ///
    /// `crucible-daemon`'s `server::fs::resolve_root` matches these two
    /// strings and refuses every other. The document says so, so the generated
    /// client offers a union rather than an open string.
    #[test]
    fn every_root_kind_is_one_the_daemon_admits() {
        use utoipa::PartialSchema;

        let table = [FsRootKind::Project, FsRootKind::Kiln];
        for kind in table {
            assert_eq!(
                serde_json::to_value(kind).expect("a kind writes JSON"),
                json!(root_kind_on_the_wire(kind)),
                "the serde spelling left the table behind"
            );
        }

        let schema = sent(FsRootKind::schema());
        let declared: Vec<&str> = schema["enum"]
            .as_array()
            .unwrap_or_else(|| panic!("`FsRootKind` is not an enum schema: {schema:#}"))
            .iter()
            .map(|value| value.as_str().expect("an enum value is a string"))
            .collect();
        let expected: Vec<&str> = table.into_iter().map(root_kind_on_the_wire).collect();
        assert_eq!(declared, expected, "the document names other kinds");
    }

    /// `SkipReason`'s wire spelling, as one exhaustive table.
    ///
    /// A fifth reason fails to compile here. The browser prints a sentence per
    /// reason, so a reason it has no arm for is a blank line in the file tree.
    fn skip_reason_on_the_wire(reason: SkipReason) -> &'static str {
        match reason {
            SkipReason::Ambiguous => "ambiguous",
            SkipReason::StaleSpan => "stale-span",
            SkipReason::CanvasNoExactMatch => "canvas-no-exact-match",
            SkipReason::CanvasUnreadable => "canvas-unreadable",
        }
    }

    /// Every skip reason keeps the spelling it had, and the document names all
    /// four.
    #[test]
    fn every_skip_reason_reaches_the_wire() {
        use utoipa::PartialSchema;

        let table = [
            SkipReason::Ambiguous,
            SkipReason::StaleSpan,
            SkipReason::CanvasNoExactMatch,
            SkipReason::CanvasUnreadable,
        ];
        for reason in table {
            assert_eq!(
                serde_json::to_value(reason).expect("a reason writes JSON"),
                json!(skip_reason_on_the_wire(reason)),
                "the serde spelling left the table behind"
            );
        }

        let schema = sent(SkipReason::schema());
        let declared: Vec<&str> = schema["enum"]
            .as_array()
            .unwrap_or_else(|| panic!("`SkipReason` is not an enum schema: {schema:#}"))
            .iter()
            .map(|value| value.as_str().expect("an enum value is a string"))
            .collect();
        let expected: Vec<&str> = table.into_iter().map(skip_reason_on_the_wire).collect();
        assert_eq!(declared, expected, "the document names other reasons");
    }
}

File: crates/crucible-web/web/src/lib/query/fs.ts
import type { Accessor } from 'solid-js';
import {
  useMutation,
  useQuery,
  type QueryClient,
  type UseMutationResult,
  type UseQueryResult,
} from '@tanstack/solid-query';
import {
  fsMkdir,
  fsMove,
  fsTrash,
  getFileWithHash,
  listDir,
  saveFileContent,
  type FsMoveOutcome,
} from '@/lib/api';
import type { FsListing } from '@/lib/types';
import { getQueryClient } from './client';
import { keys } from './keys';

/**
 * The filesystem as one cache: a listing per folder, and a file per path.
 *
 * Every entry here is held under an ABSOLUTE path, because that is the only
 * name the daemon's stream uses. `/api/fs/events` says `/kiln/notes/a.md` and
 * nothing else — no root, no relative path — so a listing held under the pair
 * the list route takes (`root`, `rel_path`) is a listing no event can reach.
 * `absFolderPath` is what turns the caller's pair into that name, and its test
 * asserts it against `folderOf`, which is what the stream's route calls.
 *
 * Six panels read files today, and each of them fetched on its own: the tree,
 * the editor, the canvas card, the backlinks snippets, the tool diff and the
 * permission prompt. Two of them looking at one note asked twice, and neither
 * learnt that the other had written to it.
 *
 * A write invalidates the keys it makes wrong ITSELF, and the stream's route
 * invalidates them again a moment later. Both is correct: the user who moved
 * the row is looking at the panel now, the daemon's event is a round trip
 * away, and a root the daemon does not watch never sends one.
 */

/** What one directory read asks for: a root, and a folder inside it. */
export interface DirRequest {
  /** The absolute root the daemon admits: a registered project or a kiln. */
  readonly root: string;
  /** POSIX path of the folder under `root`. Empty names the root itself. */
  readonly relPath?: string;
  /** Whether dotfiles are listed. NOT part of the key — see `dirOptions`. */
  readonly showHidden?: boolean;
}

/** What one move asks for. `kind` picks the daemon's allowlist. */
export interface FsMoveParams {
  readonly root: string;
  readonly kind: 'project' | 'kiln';
  readonly fromRel: string;
  readonly toRel: string;
}

/** What one folder creation or one trash asks for. */
export interface FsPathParams {
  readonly root: string;
  readonly kind: 'project' | 'kiln';
  readonly relPath: string;
}

/** What one whole-file save asks for. */
export interface SaveFileParams {
  readonly path: string;
  readonly content: string;
}

/**
 * The folder one absolute path is in.
 *
 * The stream's route calls it on the path an event names, and `absFolderPath`
 * answers the same strings, so a reader and an event meet on one key. A path
 * directly under the root answers the root itself: an empty string there would
 * name a folder nothing lists, and the event would reach no reader.
 */
export function folderOf(path: string): string {
  const cut = path.lastIndexOf('/');
  if (cut < 0) return '/';
  return cut === 0 ? '/' : path.slice(0, cut);
}

/**
 * The absolute folder one `(root, relPath)` pair names.
 *
 * The daemon's list route takes the pair and answers about the folder; the
 * daemon's stream names the folder and knows no pair. This is the one place
 * the two meet, so a caller that holds a pair keys its listing where an event
 * can find it.
 */
export function absFolderPath(root: string, relPath = ''): string {
  const base = root.replace(/\/+$/, '');
  // A caller spells the folder it is already in as `''` or as `.`, and `./x`
  // is the same folder as `x`. Left alone, a dot names `<root>/.`, which is a
  // folder the daemon never says — so the key would be one no event reaches.
  const rel = relPath.replace(/^\.(?=$|\/)/, '').replace(/^\/+|\/+$/g, '');
  if (!rel) return base || '/';
  return `${base}/${rel}`;
}

/**
 * The options of one listing.
 *
 * `showHidden` is deliberately NOT in the key. The key must be the absolute
 * folder and nothing else, or the stream's route invalidates a key no reader
 * holds. The toggle therefore invalidates the folders it changes the answer
 * for, through `invalidateDirsUnder`.
 */
function dirOptions(request: DirRequest) {
  return {
    queryKey: keys.fsDir(absFolderPath(request.root, request.relPath)),
    queryFn: () => listDir(request.root, request.relPath ?? '', request.showHidden ?? false),
  };
}

/** The options of one file's bytes. */
function fileOptions(path: string) {
  return { queryKey: keys.fsFile(path), queryFn: () => getFileWithHash(path) };
}

/**
 * One folder's entries.
 *
 * The request is an accessor because the browsed root is a control the user
 * changes while the panel is mounted, and `null` is "nothing to list yet",
 * which is a different state from an empty folder.
 */
export function useListDir(
  request: Accessor<DirRequest | null>,
): UseQueryResult<FsListing, Error> {
  return useQuery(() => {
    const asked = request();
    return {
      ...(asked ? dirOptions(asked) : { queryKey: keys.fsDir(''), queryFn: () => listDir('') }),
      enabled: asked !== null,
    };
  }, getQueryClient);
}

/**
 * One folder's entries, as a promise.
 *
 * The file tree is a RECURSION over the folders the user left expanded, so it
 * cannot mount an observer per folder — it does not know their names until the
 * level above answers. It reads the same entries under the same keys, so a
 * folder an observer already holds is not fetched again, and a folder an event
 * made wrong is.
 */
export function fetchDirOnce(request: DirRequest): Promise<FsListing> {
  return getQueryClient().fetchQuery(dirOptions(request));
}

/**
 * Makes every held listing at or under one root wrong.
 *
 * Two callers need it, and neither is a write: the hidden-files toggle, which
 * changes what the SAME folder answers, and the tree's explicit refresh. The
 * key carries the folder alone, so this walks the held keys rather than naming
 * one.
 */
export function invalidateDirsUnder(root: string): Promise<void> {
  const base = absFolderPath(root);
  return getQueryClient()
    .invalidateQueries({
      predicate: (query) => {
        const [family, kind, path] = query.queryKey as unknown[];
        if (family !== 'fs' || kind !== 'dir' || typeof path !== 'string') return false;
        return base === '/' || path === base || path.startsWith(`${base}/`);
      },
    })
    .then(() => undefined);
}

/**
 * One file's bytes and the hash of them.
 *
 * The hash is what a guarded write anchors on, so the read that fills the
 * cache is the one that carries it. `useGetFileContent` selects the text out
 * of the same entry rather than holding a second copy under the same key —
 * two shapes under one key is a reader that crashes on another reader's data.
 */
export function useGetFileWithHash(
  path: Accessor<string | null>,
): UseQueryResult<{ content: string; content_hash: string }, Error> {
  return useQuery(() => {
    const asked = path();
    return { ...fileOptions(asked ?? ''), enabled: asked !== null };
  }, getQueryClient);
}

/** One file's text. */
export function useGetFileContent(path: Accessor<string | null>): UseQueryResult<string, Error> {
  return useQuery(() => {
    const asked = path();
    return {
      ...fileOptions(asked ?? ''),
      enabled: asked !== null,
      select: (file: { content: string }) => file.content,
    };
  }, getQueryClient);
}

/**
 * One file's text, as a promise.
 *
 * For the readers that act rather than render: the tool card building a diff
 * on a click, and the backlinks panel pulling a snippet out of each linking
 * note. Both read under the key the panels observe, so the third reader of a
 * note joins the two that are already holding it.
 */
export function fetchFileContentOnce(path: string): Promise<string> {
  return getQueryClient()
    .fetchQuery(fileOptions(path))
    .then((file) => file.content);
}

/** The file, and the folder whose row carries its size and modified time. */
function invalidateFile(client: QueryClient, path: string): Promise<void> {
  return Promise.all([
    client.invalidateQueries({ queryKey: keys.fsFile(path) }),
    client.invalidateQueries({ queryKey: keys.fsDir(folderOf(path)) }),
  ]).then(() => undefined);
}

/** Every named folder, each asked for once. */
function invalidateFolders(client: QueryClient, folders: string[]): Promise<void> {
  return Promise.all(
    [...new Set(folders)].map((folder) =>
      client.invalidateQueries({ queryKey: keys.fsDir(folder) }),
    ),
  ).then(() => undefined);
}

/**
 * Every path one write moved, and the folder each of them is in.
 *
 * A move and a trash change the BYTES at a path as much as they change the
 * listing around it: the old path holds nothing now, and the new one holds
 * what was at the old one. A reader of either is holding an answer about a
 * file that is no longer there, so this names the file keys as well — the
 * same pair the stream's route names, for the same reason.
 */
function invalidatePaths(client: QueryClient, paths: string[]): Promise<void> {
  const each = [...new Set(paths)];
  return Promise.all([
    ...each.map((path) => client.invalidateQueries({ queryKey: keys.fsFile(path) })),
    invalidateFolders(client, each.map(folderOf)),
  ]).then(() => undefined);
}

/**
 * Writes a whole file, then makes the entries it changed wrong.
 *
 * The daemon writes what it is given, so this is the unguarded path: the
 * canvas card's own document and the tree's "new note". A note the USER is
 * editing goes through the offline outbox instead, which sends the hash it
 * read and lets the daemon refuse a stale write.
 *
 * It is a plain function beside the hook because one caller writes while it is
 * going away: virtualization unmounts a canvas card that leaves the viewport,
 * and that card flushes its queued edit from `onCleanup`. A write that depends
 * on a live observer would make virtualization into data loss.
 */
export function saveFileOnce(path: string, content: string): Promise<void> {
  return saveFileContent(path, content).then(() => invalidateFile(getQueryClient(), path));
}

/** The same write, for a caller that wants its pending and error state. */
export function useSaveFileContent(): UseMutationResult<void, Error, SaveFileParams> {
  return useMutation(
    () => ({
      mutationFn: ({ path, content }: SaveFileParams) => saveFileOnce(path, content),
    }),
    getQueryClient,
  );
}

/**
 * Moves or renames one path inside one root.
 *
 * Both ends are invalidated — the bytes and the listing at each — and a rename
 * inside one folder names that folder once: two invalidations of one listing
 * are two refetches for one move.
 */
export function useFsMove(): UseMutationResult<FsMoveOutcome, Error, FsMoveParams> {
  return useMutation(
    () => ({
      mutationFn: ({ root, kind, fromRel, toRel }: FsMoveParams) =>
        fsMove(root, kind, fromRel, toRel),
      onSuccess: (_outcome: FsMoveOutcome, { root, fromRel, toRel }: FsMoveParams) =>
        invalidatePaths(getQueryClient(), [
          absFolderPath(root, fromRel),
          absFolderPath(root, toRel),
        ]),
    }),
    getQueryClient,
  );
}

/** Creates a folder; the folder that gained a row is asked for again. */
export function useFsMkdir(): UseMutationResult<void, Error, FsPathParams> {
  return useMutation(
    () => ({
      mutationFn: ({ root, kind, relPath }: FsPathParams) => fsMkdir(root, kind, relPath),
      onSuccess: (_result: void, { root, relPath }: FsPathParams) =>
        invalidateFolders(getQueryClient(), [folderOf(absFolderPath(root, relPath))]),
    }),
    getQueryClient,
  );
}

/** Moves one path to the root's trash; its bytes and its folder are asked for again. */
export function useFsTrash(): UseMutationResult<void, Error, FsPathParams> {
  return useMutation(
    () => ({
      mutationFn: ({ root, kind, relPath }: FsPathParams) => fsTrash(root, kind, relPath),
      onSuccess: (_result: void, { root, relPath }: FsPathParams) =>
        invalidatePaths(getQueryClient(), [absFolderPath(root, relPath)]),
    }),
    getQueryClient,
  );
}
